File: content/snippets/python/design_patterns_singleton.py
class Config:
    _instance: "Config | None" = None

    def __new__(cls) -> "Config":
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance.env = "dev"
        return cls._instance


Config().env = "prod"
print(Config().env)

File: content/snippets/cpp/design_patterns_singleton.cpp
#include <iostream>
#include <string>

class Config {
public:
    static Config& instance() {
        static Config config;
        return config;
    }

    void setEnv(std::string value) {
        env_ = std::move(value);
    }

    const std::string& env() const {
        return env_;
    }

private:
    Config() = default;
    std::string env_ = "dev";
};

int main() {
    Config::instance().setEnv("prod");
    std::cout << Config::instance().env() << '\n';
}